src/client/main.cpp

```diff
@@ -17,59 +18,52 @@
 
 #include "frida/frida-gum.h"
 
-#define FB_ID "mxcfb"
-
-#ifndef _GNU_SOURCE
-#define _GNU_SOURCE
-#endif
-
-int msg_q_id = 0x2257c;
+constexpr auto msg_q_id = 0x2257c;
 swtfb::ipc::Queue MSGQ(msg_q_id);
 
-uint16_t *SHARED_BUF;
+uint16_t *SHARED_BUF = nullptr;
 
-const int BYTES_PER_PIXEL = sizeof(uint16_t);
+constexpr auto BYTES_PER_PIXEL = sizeof(*SHARED_BUF);
 
 bool IN_XOCHITL = false;
+bool ON_RM2 = false;
 
 extern "C" {
-struct stat;
 
-// QImage(width, height, format)
-static void (*qImageCtor)(void *that, int x, int y, int f) = 0;
-// QImage(uchar*, width, height, bytesperline, format)
-static void (*qImageCtorWithBuffer)(void *that, uint8_t *, int32_t x, int32_t y,
-                                    int32_t bytes, int format, void (*)(void *),
-                                    void *) = 0;
-
-static void _libhook_init() __attribute__((constructor, visibility("default")));
-static void _libhook_init() {
+__attribute__((constructor))
+void init() {
   std::ios_base::Init i;
 
-  auto VERSION = "0.1";
+  constexpr auto VERSION = "0.1";
 
   setenv("RM2FB_SHIM", VERSION, true);
 
-  if (getenv("RM2FB_ACTIVE") != "") {
+  if (getenv("RM2FB_ACTIVE") != nullptr) {
     setenv("RM2FB_NESTED", "1", true);
   } else {
     setenv("RM2FB_ACTIVE", "1", true);
   }
```

Comment:
these should only be set on rM2 (not on rM1)

Reply:
Thanks for catching that! Should be fixed now. I tested it by running the program below with `LD_PRELOAD=./librm2fb_client.so.1.0.0` and checking its output on rM1 and rM2.

<details>
<summary>Test program</summary>

```c
#include <stdio.h>
#include <stdlib.h>
#include <sys/wait.h>
#include <unistd.h>

int main(int argc, char** argv) {
    if (argc == 1) {
        printf("Parent\n");
    }

    printf("RM2FB_SHIM = %s\n", getenv("RM2FB_SHIM"));
    printf("RM2FB_ACTIVE = %s\n", getenv("RM2FB_ACTIVE"));
    printf("RM2FB_NESTED = %s\n", getenv("RM2FB_NESTED"));

    if (argc == 1) {
        if (fork() == 0) {
            printf("\nChild\n");
            execl("./test_env", "./test_env", "stop");
        } else {
            wait(NULL);
        }
    }
}
```
</details>

Output on rM1:

```
Parent
RM2FB_SHIM = (null)
RM2FB_ACTIVE = (null)
RM2FB_NESTED = (null)

Child
RM2FB_SHIM = (null)
RM2FB_ACTIVE = (null)
RM2FB_NESTED = (null)
```

Output on rM2:

```
OPENED SHARED MEM: /dev/shm/swtfb.01 at 75f46000, errno: 0
Parent
RM2FB_SHIM = 0.1
RM2FB_ACTIVE = 1
RM2FB_NESTED = (null)

Child
RM2FB_SHIM = 0.1
RM2FB_ACTIVE = 1
RM2FB_NESTED = 1
```